File: 110BalancedBinaryTree.cpp
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return diffHeigh(root)>=0; 
    }
    
    int diffHeigh(TreeNode* root){
        if(!root) return 0;
        
        int left = diffHeigh(root->left);
        int right = diffHeigh(root->right);
        
        if(left<0 || right<0 || abs(left-right)>1) return -1;  //If left or right has unbalanced component, return -1 anyway.
        
        return max(left,right)+1;
    }
};

File: Algo/Leetcode/110BalancedBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def __init__(self):
        self.res = True

    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True

        self.dfs(root)

        return self.res

    def dfs(self, root):
        if root is None:
            return 0

        d_l = self.dfs(root.left)
        d_r = self.dfs(root.right)

        if abs(d_l - d_r) > 1:
            self.res = False

        return max(d_r, d_l) + 1